src/stream/src/executor/source/source_backfill_executor.rs

```diff
@@ -414,24 +421,39 @@ impl<S: StateStore> SourceBackfillExecutorInner<S> {
 
         type PausedReader = Option<impl Stream>;
         let mut paused_reader: PausedReader = None;
-        let mut command_paused = false;
 
         macro_rules! pause_reader {
             () => {
-                let (left, right) = backfill_stream.into_inner();
-                backfill_stream = select_with_strategy(
-                    left,
-                    futures::stream::pending().boxed().map(Either::Right),
-                    select_strategy,
-                );
-                // XXX: do we have to store the original reader? Can we simply rebuild the reader later?
-                paused_reader = Some(right);
+                if !pause_control.reader_paused {
+                    let (left, right) = backfill_stream.into_inner();
+                    backfill_stream = select_with_strategy(
+                        left,
+                        futures::stream::pending().boxed().map(Either::Right),
+                        select_strategy,
+                    );
+                    // XXX: do we have to store the original reader? Can we simply rebuild the reader later?
+                    paused_reader = Some(right);
+                    pause_control.reader_paused = true;
+                }
+            };
+        }
+
+        macro_rules! resume_reader {
+            () => {
+                if pause_control.reader_paused {
+                    backfill_stream = select_with_strategy(
+                        input.by_ref().map(Either::Left),
+                        paused_reader
+                            .take()
+                            .expect("should have paused reader to resume"),
+                        select_strategy,
+                    );
+                    pause_control.reader_paused = false;
+                }
             };
         }
 
-        // If the first barrier requires us to pause on startup, pause the stream.
-        if is_pause_on_startup {
-            command_paused = true;
+        if pause_control.pause_on_startup() {
             pause_reader!();
         }
```

Comment:
I think the semantic of the `fn pause_on_startup` is not ideal.

Either of the below might be better:
1. `fn paused(&self) { self.backfill_paused || self.command_paused || self.self_paused }`
2. just still use a local variable `is_pause_on_startup`